### ISSUE TITLE
The performance is not particularly great

### ISSUE THREAD

**Issue by vtereshkov** [vtereshkov/umka-lang#163]
Take something like this
```go
fn main() {
    x := 0.0
    for i := 0.0; i <= 10000000.0; i += 1 {
        x += i
        x /= 2.0
    }
    printf("%f\n", x);
}
```
versus this
```lua
x = 0
for i=0,10000000 do
    x = x + i
    x = x / 2
end

print(x)
```
First is umka, second is lua. On my computer at least, umka is about 3x slower. Which is a shame because it has potential to be much faster than lua. I usually don't pay attention to the speed of scripting languages, since it's almost always not the issue, however since I'm writing almost the entire game in umka, the speed matters, and it shows: Most of the "self" execution time is spent in vmRun. 
The above isn't a perfect benchmark, so here's few more (these benchmarks mostly tackle umka's "raw" speed, e.g. crunching speed):

```go
fn fib(x: int): int {
    if x < 2 {
        return x
    }
    return fib(x-1) + fib(x-2)
}

fn main() {
    printf("%d", fib(40))
}
```
20 seconds
vs
```lua
x = 0
for i=0,10000000 do
    x = x + i
    x = x / 2
end

print(x)
```
10 seconds
(lua about 2x faster)

---

```go
type (
    StringMod = interface {
        f(s: str): str
    }

    TypeA = struct {}
    TypeB = struct {}
)

fn (t: ^TypeA) f(s: str): str {
    return s + "a"
}

fn (t: ^TypeB) f(s: str): str {
    return "b"
}

fn apply(iface: StringMod, n: int): str {
    s := ""
    for i := 0; i <= n; i++ {
        s = iface.f(s)
    }
    return s
}

fn main() {
    a := TypeA{}
    b := TypeB{}
    apply(a, 100000)
    apply(b, 100000)
}
```
(interface version) 600ms
```go
fn apply(f: fn (s: str): str, n: int): str {
    s := ""
    for i := 0; i <= n; i++ {
        s = f(s)
    }
    return s
}

fn main() {
    apply(fn (s: str): str { return s + "a"; }, 100000)
    apply(fn (s: str): str { return "b" }, 100000)
}
```
(function pointer version) 900ms
```lua
function apply(interface, n)
    s = ''
    for i=0,n do 
        s = interface.f(s)
    end
    return s
end

a = { f = function(s) return s .. 'a' end }
b = { f = function(s) return 'b' end }

apply(a, 100000)
apply(b, 100000)
```
("interface" version) 400ms.
```lua
function apply(f, n)
    s = ''
    for i=0,n do 
        s = f(s)
    end
    return s
end

apply(function(s) return s .. 'a' end, 100000)
apply(function(s) return 'b' end, 100000)
```
(function pointer version) 566ms 

There's more I can show. But I just think there's room for improvement when it comes to performance. I've tried to re-implement umka with focus of VM performance, but the project is nowhere near as complete as umka. While it's faster than Lua in general, I don't think it's fair to compare it with umka yet. (https://github.com/skejeton/elka), It's pretty much abandoned//

Either way, good luck. And if something, just ask me. I'll be glad to respond and/or help.

_Originally posted by @skejeton in https://github.com/marekmaskarinec/tophat/issues/30#issuecomment-1158684630_

**Comment by vtereshkov** [comment]
Related issue: #140

**Comment by vtereshkov** [comment]
@skejeton Your Lua example:
```
x = 0
for i=0,10000000 do
    x = x + i
    x = x / 2
end

print(x)
```
can be even 10x faster than Umka if you declare `x` as `local x = 0`, so that it can be treated as a conventional variable rather than an upvalue.

Sad for Umka.

**Comment by marekmaskarinec** [comment]
Perhaps we could benefit from adding more complex instructions? For example a variant of binary, which works on the top two elements of the stack. Would that be worth it?

**Comment by vtereshkov** [comment]
@luauser32167 You're right, the Umka virtual machine is not optimized enough. Please notice, however, that a one-to-one comparison between Umka and Lua 5 bytecode is hardly possible, since Umka is stack-based and Lua 5 is register-based.

@marekmaskarinec I sometimes think of having a single complex instruction for a `for` loop header in its most widely used form: `for i := 0; i < expr; i++`. AFAIK, even Lua 4 featured it.
